src/devices/machine/mc6844.cpp

```diff
@@ -0,0 +1,556 @@
+	: device_t(mconfig, MC6844, tag, owner, clock)
+	, device_execute_interface(mconfig, *this)
+	, m_out_int_cb(*this)
+	, m_out_txak_cb(*this)
+	, m_out_drq1_cb(*this)
+	, m_out_drq2_cb(*this)
+	, m_in_memr_cb(*this)
+	, m_out_memw_cb(*this)
+	, m_in_ior_cb{ { *this },{ *this },{ *this },{ *this } }
+	, m_out_iow_cb{ { *this },{ *this },{ *this },{ *this } }
+	, m_state(STATE_S0)
+	, m_icount(0)
+{
+}
+
+//-------------------------------------------------
+//  device_resolve_objects - device-specific setup
+//-------------------------------------------------
+void mc6844_device::device_resolve_objects()
+{
+	m_out_int_cb.resolve_safe();
+	m_out_txak_cb.resolve_safe();
+	m_out_drq1_cb.resolve_safe();
+	m_out_drq2_cb.resolve_safe();
+	m_in_memr_cb.resolve_safe(0);
+	m_out_memw_cb.resolve_safe();
+
+	for(auto &cb : m_in_ior_cb)
+			cb.resolve_safe(0);
+	for(auto &cb : m_out_iow_cb)
+			cb.resolve_safe();
+
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void mc6844_device::device_start()
+{
+	// set our instruction counter
+	set_icountptr(m_icount);
+	
+	save_item(NAME(m_m6844_priority));
+	save_item(NAME(m_m6844_interrupt));
+	save_item(NAME(m_m6844_chain));
+	save_item(NAME(m_state));
+	save_item(NAME(m_icount));
+	save_item(NAME(m_current_channel));
+	save_item(NAME(m_last_channel));
+	save_item(NAME(m_dgrnt));
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void mc6844_device::device_reset()
+{
+	// reset the 6844
+	for (int i = 0; i < 4; i++)
+	{
+		m_m6844_channel[i].active = 0;
+		m_m6844_channel[i].control = 0x00;
+	}
+	m_m6844_priority = 0x00;
+	m_m6844_interrupt = 0x00;
+	m_m6844_chain = 0x00;
+	m_state = STATE_SI;
+}
+
+//-------------------------------------------------
+//  dma_request -
+//-------------------------------------------------
+
+void mc6844_device::dma_request(int channel, int state)
+{
+	LOG("MC6844 Channel %u DMA Request: %u\n", channel, state);
+
+	m_dreq[channel & 3] = state;
+
+	LOGSTATE("Trigger(1)\n");
+	trigger(1);
+}
+
+//-------------------------------------------------
+//  execute_run -
+//-------------------------------------------------
+
+void mc6844_device::execute_run()
+{
+	do
+	{
+		switch (m_state)
+		{
+		case STATE_SI:  // IDLE state, will suspend until a DMA request comes through
+			{   //                    Hi ------> Lo
+				int const priorities[][4] = {{ 1, 2, 3, 0 },
+							   { 2, 3, 0, 1 },
+							   { 3, 0, 1, 2 },
+							   { 0, 1, 2, 3 }};
+
+				LOGSTATE("DMA state SI\n");
+				for (int prio = 0; prio < 4; prio++)
+				{
+					// Rotating or static channel prioritizations
+					int current_channel = priorities[((m_m6844_priority & 0x80) ? m_last_channel : 3) & 3][prio];
+
+				if (m_m6844_channel[current_channel].active == 1 && m_dreq[current_channel] == ASSERT_LINE)
```

Comment:
Why is this line outdented?

Reply:
Hmm, I ran srcclean before I committed, thought those things were taken care of then, apparently it caused some new indentation level in TAB translations, I'll go over and fix it